### PR TITLE
Remembering Notifications Reply Text

### LINKED ISSUE #6003: Notifications: Remember Reply Text
Improve ReplyTextView Behavior: Remember any input after hitting Back!

### PR DESCRIPTION
**Pull request by jleandroperez**
### Details:
This PR implements a Notification Reply cache, which preserved any entered text for 7 days.

Closes #6003

### To test:
1. Verify the (new) unit tests are green
2. Launch WPiOS and log into your account
3. Open any Comment-Y Notification
4. Press over the ReplyTextView component
5. Enter some random text
6. Hit back
7. Kill the app, and relaunch
8. Reopen the same notofication opened in (3)

Verify that the Reply Text is pre-populated with the text you've written in (5).

Needs Review: @SergioEstevao 
Sir, may i bug you with a fun review?

Thanks in advance!!

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Notifications/Tools/NotificationReplyStore.swift [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    /// Our beautiful and private Initializer. In here we'll trigger the cleanup sequence, which removes outdated replies!.
+    ///
+    private init() {
+        purgeOldReplies()
+    }
+
+    /// Unit Testing Helper: Allows us to hack the current date.
+    ///
+    init(now: Date) {
```
Comment (SergioEstevao) [inline WordPress/Classes/ViewRelated/Notifications/Tools/NotificationReplyStore.swift]: You probably can mark this private also? and then in the tests say that you want access to the private methods.

**Comment by SergioEstevao** [comment]
I'm seeing a crash when I'm doing steps 7 to 8. I relaunch the app, tap on the Notification tab and see this crash:

```2018-03-06 10:38:09.714370+0000 WordPress[6387:80638] Fatal error: Unexpectedly found nil while unwrapping an Optional value```

It's on this method of NotificationDetailsViewController:

```func storeNotificationReplyIfNeeded() {
        guard let reply = replyTextView.text, !reply.isEmpty else {
            return
        }

        NotificationReplyStore.shared.store(reply: reply, for: note.notificationId)
    }```

**Comment by jleandroperez** [comment]
@SergioEstevao NICE CATCH on the crash!!!. Problem is!!! can't trigger that, on my end.

I've added safe unwraps *all over* the snippet that was crashing. Mind taking this for another pass?

Thanks!!!

File: WordPress/Classes/ViewRelated/Notifications/Tools/NotificationReplyStore.swift (hunk 1 above)
Comment (jleandroperez) [reply WordPress/Classes/ViewRelated/Notifications/Tools/NotificationReplyStore.swift]: I would absolutely love to do so. But the `@testable import` trick, in Unit Tests, only lets you call *internal* methods, not private.

I really really hope Swift 5 comes with annotations!!! thanks for the suggestion though!!

**Comment by jleandroperez** [comment]
Thank you Sergio!!

**Pull request merged**